pylint/lint/pylinter.py

```diff
@@ -641,6 +641,13 @@ def check(self, files_or_modules: Sequence[str] | str) -> None:
 
             filepath = files_or_modules[0]
             with fix_import_path(files_or_modules):
+                if _is_ignored_file(
```

Comment:
Shouldn't we also remove the previous place where we where ignoring some file ? If this is the right place then we don't need to check in the other places it was checked ?

Reply:
I am not completely sure I understand. The check was not present previously in this if statement. Do you mean that we should put the check before the second if statement and remove it from all the subsequent if statements?

Reply:
We're already filtering files somewhere when they are not from stdin, I think there's a kind of refactor to do there so we filter files at the right place and so it's efficient (done only once per run, possibly when using multiprocessing). I would start by checking where ``_is_ignored_file`` is used.

Reply:
Okay, I looked into it a bit more and it seems we are filtering files while doing the following tasks:

Reply:
That function is used in two places:
* [When disovering files recursively](https://github.com/PyCQA/pylint/blob/8ccb52650f78f0650081753b218375bfcf47ab64/pylint/lint/pylinter.py#L600)
* [When expanding modules](https://github.com/PyCQA/pylint/blob/8ccb52650f78f0650081753b218375bfcf47ab64/pylint/lint/expand_modules.py#L79)

It seems like the `expand_modules` function is the place where the filtering of ignored files is meant to take place and the use of `_is_ignored_file` while discovering is more of a "hack". Unfortunately `expand_modules` does not get called when linting from standard input. Furthermore `expand_modules` seems to do two things:
1. Expand modules
2. Filter ignored files

It might be a good idea to entangle these two concerns. In any case I don't see a way to only filter once. Before `expand_modules` you want to filter once to avoid unnecessarily processing any ignored files but you can not filter everything at this stage because you do not have filepaths for the modules yet. Therefore you want to filter again after you have those.

So I am not sure how to proceed from here.

Reply:
Thank you for checking this.

> It might be a good idea to entangle these two concerns. In any case I don't see a way to only filter once. Before expand_modules you want to filter once to avoid unnecessarily processing any ignored files but you can not filter everything at this stage because you do not have filepaths for the modules yet. Therefore you want to filter again after you have those.

Hmm, it sounds like a bigger refactor than I thought. If we're going to extract the file discovering from PyLinter we could expand modules, reading from stdin and files discovery at the same time, right ? Removing concerns from the PyLinter would be a good thing especially for more efficient multiprocessing (easier to map files to threads), but I don't have a clear vision about the issue we'll face when doing that.

Reply:
Yes, I think it should be possible to do the discovering, filtering and expanding beforehand and then passing the result to `Pylinter.check` but as you said that is going to be a larger refactor.

Reply:
Would you be willing to give it a try ?

Reply:
I would like to give it a try. That said I have no idea how long it will take because it seems like a quite big refactor and I do not have a lot of time to work on this.

Reply:
Nice to hear, don't worry about deadline there is none :) ! You can open a draft pull request to discuss the refactor with maintainers, early feedback would probably help you.